store/layered_store.go

```diff
@@ -0,0 +1,186 @@
+		DatabaseLayer: NewSqlSupplier(),
+	}
+}
+
+type QueryFunction func(LayeredStoreSupplier) LayeredStoreSupplierResult
+
+func (s *LayeredStore) RunQuery(queryFunction QueryFunction) StoreChannel {
+	storeChannel := make(StoreChannel)
+
+	go func() {
+		finalResult := StoreResult{}
+		// Logic for determining what layers to run
+		if result := queryFunction(s.DatabaseLayer); result.Err == nil {
+			finalResult.Data = result.Result
+		} else {
+			finalResult.Err = result.Err
+		}
+
+		storeChannel <- finalResult
+	}()
+
+	return storeChannel
+}
+
+func (s *LayeredStore) Team() TeamStore {
+	return s.DatabaseLayer.Team()
+}
+
+func (s *LayeredStore) Channel() ChannelStore {
+	return s.DatabaseLayer.Channel()
+}
+
+func (s *LayeredStore) Post() PostStore {
+	return s.DatabaseLayer.Post()
+}
+
+func (s *LayeredStore) User() UserStore {
+	return s.DatabaseLayer.User()
+}
+
+func (s *LayeredStore) Audit() AuditStore {
+	return s.DatabaseLayer.Audit()
+}
+
+func (s *LayeredStore) ClusterDiscovery() ClusterDiscoveryStore {
+	return s.DatabaseLayer.ClusterDiscovery()
+}
+
+func (s *LayeredStore) Compliance() ComplianceStore {
+	return s.DatabaseLayer.Compliance()
+}
+
+func (s *LayeredStore) Session() SessionStore {
+	return s.DatabaseLayer.Session()
+}
+
+func (s *LayeredStore) OAuth() OAuthStore {
+	return s.DatabaseLayer.OAuth()
+}
+
+func (s *LayeredStore) System() SystemStore {
+	return s.DatabaseLayer.System()
+}
+
+func (s *LayeredStore) Webhook() WebhookStore {
+	return s.DatabaseLayer.Webhook()
+}
+
+func (s *LayeredStore) Command() CommandStore {
+	return s.DatabaseLayer.Command()
+}
+
+func (s *LayeredStore) Preference() PreferenceStore {
+	return s.DatabaseLayer.Preference()
+}
+
+func (s *LayeredStore) License() LicenseStore {
+	return s.DatabaseLayer.License()
+}
+
+func (s *LayeredStore) Token() TokenStore {
+	return s.DatabaseLayer.Token()
+}
+
+func (s *LayeredStore) Emoji() EmojiStore {
+	return s.DatabaseLayer.Emoji()
+}
+
+func (s *LayeredStore) Status() StatusStore {
+	return s.DatabaseLayer.Status()
+}
+
+func (s *LayeredStore) FileInfo() FileInfoStore {
+	return s.DatabaseLayer.FileInfo()
+}
+
+func (s *LayeredStore) Reaction() ReactionStore {
+	return s.DatabaseLayer.Reaction()
+}
+
+func (s *LayeredStore) JobStatus() JobStatusStore {
+	return s.DatabaseLayer.JobStatus()
+}
+
+func (s *LayeredStore) MarkSystemRanUnitTests() {
+	s.DatabaseLayer.MarkSystemRanUnitTests()
+}
+
+func (s *LayeredStore) Close() {
+	s.DatabaseLayer.Close()
+}
+
+func (s *LayeredStore) DropAllTables() {
+	s.DatabaseLayer.DropAllTables()
+}
+
+func (s *LayeredStore) TotalMasterDbConnections() int {
+	return s.DatabaseLayer.TotalMasterDbConnections()
+}
+
+func (s *LayeredStore) TotalReadDbConnections() int {
+	return s.DatabaseLayer.TotalReadDbConnections()
+}
+
+func (s *LayeredStore) TotalSearchDbConnections() int {
+	return s.DatabaseLayer.TotalSearchDbConnections()
+}
+
+type LayeredReactionStore struct {
+	*LayeredStore
+}
```

Comment:
Ok, just worried about the file getting reallllly long once we have all the other stores in here as well